pallets/loans/src/lib.rs

```diff
@@ -699,6 +737,439 @@ impl<T: Config> Pallet<T> {
+            .ok_or(Error::<T>::CollateralOverflow)?;
+
+        currency_price
+            .checked_mul(&FixedU128::from_inner(collateral_amount))
+            .ok_or(Error::<T>::CollateralOverflow)
+    }
+
+    fn total_collateral_asset_value(
+        borrower: &T::AccountId,
+    ) -> result::Result<FixedU128, Error<T>> {
+        let collateral_assets = AccountCollateralAssets::<T>::get(borrower);
+        if collateral_assets.is_empty() {
+            return Err(Error::<T>::NoCollateralAsset);
+        }
+
+        let mut total_asset_value: FixedU128 = FixedU128::zero();
+        for currency_id in collateral_assets.iter() {
+            total_asset_value = total_asset_value
+                .checked_add(&Self::collateral_asset_value(borrower, currency_id)?)
+                .ok_or(Error::<T>::CollateralOverflow)?;
+        }
+
+        Ok(total_asset_value)
+    }
+
+    /// Borrower shouldn't borrow more than what he/she has pledged in total
+    fn borrow_guard(
+        borrower: &T::AccountId,
+        borrow_currency_id: &CurrencyId,
+        borrow_amount: Balance,
+    ) -> DispatchResult {
+        let total_will_borrow_value =
+            Self::total_will_borrow_value(borrower, borrow_currency_id, borrow_amount)?;
+        let total_collateral_asset_value = Self::total_collateral_asset_value(borrower)?;
+
+        if total_collateral_asset_value < total_will_borrow_value {
+            return Err(Error::<T>::InsufficientCollateral.into());
+        }
+
+        Ok(())
+    }
+
+    #[transactional]
+    fn repay_borrow_internal(
+        borrower: &T::AccountId,
+        currency_id: &CurrencyId,
+        repay_amount: Balance,
+    ) -> DispatchResult {
+        let account_borrows = Self::borrow_balance_stored(borrower, currency_id)?;
+        if account_borrows < repay_amount {
+            return Err(Error::<T>::RepayAmountTooBig.into());
+        }
+
+        T::Currency::transfer(*currency_id, borrower, &Self::account_id(), repay_amount)?;
+
+        let account_borrows_new = account_borrows
+            .checked_sub(repay_amount)
+            .ok_or(Error::<T>::CalcBorrowBalanceFailed)?;
+        let total_borrows = Self::total_borrows(currency_id);
+        let total_borrows_new = total_borrows
+            .checked_sub(repay_amount)
+            .ok_or(Error::<T>::CalcBorrowBalanceFailed)?;
+
+        AccountBorrows::<T>::insert(
+            currency_id,
+            borrower,
+            BorrowSnapshot {
+                principal: account_borrows_new,
+                borrow_index: Self::borrow_index(currency_id),
+            },
+        );
+
+        TotalBorrows::<T>::insert(currency_id, total_borrows_new);
+
+        Ok(())
+    }
+
+    fn collateral_asset_internal(
+        who: T::AccountId,
+        currency_id: CurrencyId,
+        enable: bool,
+    ) -> result::Result<(), Error<T>> {
+        let mut collateral_assets = AccountCollateralAssets::<T>::get(&who);
+        if enable {
```

Comment:
return error early to remove nested if else.

Reply:
I will refactor this after merge `AccountCollateralAssets` into `AccountCollateral`.